### PR TITLE
Add Reflex.Process

### PR DESCRIPTION
**Pull request by ali-abrar**
Run shell commands from your reflex app and interact with them using an FRP interface.

### CODE REVIEW

File: src/Reflex/Process.hs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        ecTrigger ec
+        P.cleanupProcess po
+        killThread outThread
+        killThread errThread 
+      return $ Process
+        { _process_exit = ecOut
+        , _process_stdout = out
+        , _process_stderr = err
+        }
+    _ -> error "Reflex.Vty.Process.createProcess: Created pipes were not returned by System.Process.createProcess."
```
Comment (alexfmpe) [inline src/Reflex/Process.hs]:
```suggestion
    _ -> error "Reflex.Vty.Process.createRedirectedProcess: Created pipes were not returned by System.Process.createProcess."
```

File: src/Reflex/Process.hs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            t <- liftIO $ forkIO reader
+            return (e, t)
+      (out, outThread) <- output hOut
+      (err, errThread) <- output hErr
+      (ecOut, ecTrigger) <- newTriggerEvent
+      void $ liftIO $ forkIO $ waitForProcess ph >>= \ec -> do
+        ecTrigger ec
+        P.cleanupProcess po
+        killThread outThread
+        killThread errThread
```
Comment (3noch) [inline src/Reflex/Process.hs]: Seems like some async exception masking would be best here

File: src/Reflex/Process.hs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import qualified System.Process as P
+import System.Process hiding (createProcess)
+
+import Reflex
+
+-- | The output of a process
+data Process t = Process
+  { _process_exit :: Event t ExitCode
+  , _process_stdout :: Event t ByteString
+  , _process_stderr :: Event t ByteString
```
Comment (3noch) [inline src/Reflex/Process.hs]: What does a `ByteString` here mean? Is it a line of input/output? Or some other buffering technique? Or is it the entire output when the process is done?

Comment (3noch) [reply src/Reflex/Process.hs]: Or I suppose there is no buffering technique on the reading side of this and it just dumps data as it comes out of the handle.

Comment (ali-abrar) [reply src/Reflex/Process.hs]: You're right: there's no buffering. We just read whatever's available and fire and event with that data, and continue doing that until the handle is closed or we see EOF. The reader loop blocks when it's waiting on input.

File: src/Reflex/Process.hs (hunk 2 above)
Comment (ali-abrar) [reply src/Reflex/Process.hs]: wrapped it in `mask_`

File: src/Reflex/Process.hs (hunk 3 above)
Comment (ali-abrar) [reply src/Reflex/Process.hs]: I switched to linebuffering on the output side

File: reflex-vty.cabal [hunk 4]
```diff
@@ -34,21 +34,25 @@ library
...
-    stm                               >= 2.4 && < 2.6,
     data-default                      >= 0.7.1 && < 0.8,
     dependent-map                     >= 0.2.4 && < 0.4,
-    text                              >= 1.2.3 && < 1.3,
-    text-icu                          >= 0.7 && < 0.8,
     dependent-sum                     >= 0.3 && < 0.7,
-    exception-transformers            >= 0.4.0 && < 0.5,
+    mtl                               >= 2.2.2 && < 2.3,
     primitive                         >= 0.6.3 && < 0.7,
+    process                           >= 1.6.4 && < 1.7,
```
Comment (Ericson2314) [inline reflex-vty.cabal]: is `process` needed in `reflex-vty` itself?

**Comment by ali-abrar** [comment]
this has been split into reflex-process, reflex-fsnotify, and reflex-ghci

**Pull request closed without merging**